app/controllers/AgentController.scala

```diff
@@ -0,0 +1,132 @@
+ * limitations under the License.
+ */
+
+package controllers
+
+import config.FrontendAppConfig
+import controllers.actions.AgentIdentifierAction.VerifyAgentClientPredicate
+import controllers.actions.{AgentIdentifierAction, DataRequiredAction, DataRetrievalAction, FeatureFlagActionFactory}
+import form.AgentClientPillar2ReferenceFormProvider
+import models.InternalIssueError
+import pages.agent.{AgentClientOrganisationNamePage, AgentClientPillar2ReferencePage}
+import play.api.i18n.I18nSupport
+import play.api.mvc.{Action, AnyContent, MessagesControllerComponents}
+import repositories.SessionRepository
+import services.SubscriptionService
+import uk.gov.hmrc.play.bootstrap.frontend.controller.FrontendBaseController
+import views.html.agent._
+import cats.implicits._
+
+import javax.inject.Inject
+import scala.concurrent.{ExecutionContext, Future}
+
+class AgentController @Inject() (
+  val controllerComponents:    MessagesControllerComponents,
+  val sessionRepository:       SessionRepository,
+  subscriptionService:         SubscriptionService,
+  clientPillarIdView:          AgentClientPillarIdView,
+  clientConfirmView:           AgentClientConfirmDetailsView,
+  clientNoMatchView:           AgentClientNoMatch,
+  agentErrorView:              AgentErrorView,
+  agentClientUnauthorisedView: AgentClientUnauthorisedView,
+  agentIndividualErrorView:    AgentIndividualErrorView,
+  agentOrganisationErrorView:  AgentOrganisationErrorView,
+  identify:                    AgentIdentifierAction,
+  featureAction:               FeatureFlagActionFactory,
+  getData:                     DataRetrievalAction,
+  requireData:                 DataRequiredAction,
+  formProvider:                AgentClientPillar2ReferenceFormProvider
+)(implicit appConfig:          FrontendAppConfig, ec: ExecutionContext)
+    extends FrontendBaseController
+    with I18nSupport {
+
+  import identify._
+
+  val form = formProvider()
+
+  def onPageLoadClientPillarId: Action[AnyContent] =
+    (featureAction.asaAccessAction andThen agentIdentify() andThen getData andThen requireData).async { implicit request =>
+      val preparedForm = request.userAnswers.get(AgentClientPillar2ReferencePage) match {
+        case Some(value) => form.fill(value)
+        case None        => form
+      }
+
+      Future.successful(Ok(clientPillarIdView(preparedForm)))
+    }
+
+  def onSubmitClientPillarId: Action[AnyContent] = (featureAction.asaAccessAction andThen agentIdentify() andThen getData andThen requireData).async {
+    implicit request =>
+      form
+        .bindFromRequest()
+        .fold(
+          formWithErrors => Future.successful(BadRequest(clientPillarIdView(formWithErrors))),
+          value => {
+            val result = for {
+              updatedAnswers     <- Future.fromTry(request.userAnswers.set(AgentClientPillar2ReferencePage, value))
+              _                  <- sessionRepository.set(updatedAnswers)
```

Comment:
Isn't it better to just do 
```suggestion
              subscriptionData   <- subscriptionService.readSubscription(value)
              answersWithOrgName <- Future.fromTry(updatedAnswers.set(AgentClientOrganisationNamePage, subscriptionData.upeDetails.organisationName))
                            _                  <- sessionRepository.set(answersWithOrgName)
```
? Not sure why you are calling set twice

Reply:
Not sure what you mean by calling set twice?

1. The first set is saving the pillar2 id the agent entered in the form
2. We use that ID to call read subscription to get the organisation name and also verify that there is an record for that ID (If this fails the agent is redirected to a KB page where they can go back and re-enter a new ID or alter the one they entered for instance they had a typo) 
3. If successful we want to save the organisation name and redirect them to the next page in the journey